src/test.rs

```diff
@@ -224,85 +224,34 @@ where
     f(&rustup_home)
 }
 
-#[cfg(feature = "otel")]
-use once_cell::sync::Lazy;
+pub async fn before_test_async() -> Option<tracing::dispatcher::DefaultGuard> {
+    #[cfg(feature = "otel")]
+    {
+        use tracing_subscriber::{layer::SubscriberExt, Registry};
 
-/// A tokio runtime for the sync tests, permitting the use of tracing. This is
-/// never shutdown, instead it is just dropped at end of process.
-#[cfg(feature = "otel")]
-static TRACE_RUNTIME: Lazy<tokio::runtime::Runtime> =
-    Lazy::new(|| tokio::runtime::Runtime::new().unwrap());
-/// A tracer for the tests.
-#[cfg(feature = "otel")]
-static TRACER: Lazy<opentelemetry_sdk::trace::Tracer> = Lazy::new(|| {
-    use std::time::Duration;
+        let telemetry = {
+            use crate::cli::log::{telemetry, telemetry_default_tracer};
 
-    use opentelemetry::{global, KeyValue};
-    use opentelemetry_otlp::WithExportConfig;
-    use opentelemetry_sdk::{
-        propagation::TraceContextPropagator,
-        trace::{self, Sampler},
-        Resource,
-    };
-    use tokio::runtime::Handle;
-    use tracing_subscriber::{layer::SubscriberExt, EnvFilter, Registry};
-
-    // Use the current runtime, or the sync test runtime otherwise.
-    let handle = match Handle::try_current() {
-        Ok(handle) => handle,
-        Err(_) => TRACE_RUNTIME.handle().clone(),
-    };
-    let _guard = handle.enter();
+            use opentelemetry::global;
+            use opentelemetry_sdk::propagation::TraceContextPropagator;
 
-    let tracer = opentelemetry_otlp::new_pipeline()
-        .tracing()
-        .with_exporter(
-            opentelemetry_otlp::new_exporter()
-                .tonic()
-                .with_timeout(Duration::from_secs(3)),
-        )
-        .with_trace_config(
-            trace::config()
-                .with_sampler(Sampler::AlwaysOn)
-                .with_resource(Resource::new(vec![KeyValue::new("service.name", "rustup")])),
-        )
-        .install_batch(opentelemetry_sdk::runtime::Tokio)
-        .unwrap();
+            global::set_text_map_propagator(TraceContextPropagator::new());
+            telemetry(telemetry_default_tracer().unwrap())
```

Comment:
Maybe this can be inlined into `telemetry().unwrap()` or even a lazy static `TELEMETRY`?

Reply:
The latter is not possible, since the output type of `telemetry()` must be generic `impl Layer<S>`.